src/common/librlist/rlist.c

```diff
@@ -1271,17 +1435,117 @@ static json_t *rlist_json_nodelist (struct rlist *rl)
     return o;
 }
 
+static zhashx_t *rlist_properties (struct rlist *rl)
+{
+    int saved_errno;
+    struct rnode *n;
+    zlistx_t *keys = NULL;
+    zhashx_t *properties = zhashx_new ();
+
+    if (!properties)
+        return NULL;
+
+    zhashx_set_destructor (properties, property_destructor);
+
+    n = zlistx_first (rl->nodes);
+    while (n) {
+        if (n->properties) {
+            const char *name;
+            if (!(keys = zhashx_keys (n->properties)))
+                goto error;
+            name = zlistx_first (keys);
+            while (name) {
+                struct idset *ids = zhashx_lookup (properties, name);
+                if (!ids) {
+                    if (!(ids = idset_create (0, IDSET_FLAG_AUTOGROW)))
+                        goto error;
+                    if (zhashx_insert (properties, name, ids) < 0) {
+                        idset_destroy (ids);
+                        goto error;
+                    }
+                }
+                if (idset_set (ids, n->rank) < 0)
+                    goto error;
+                name = zlistx_next (keys);
+            }
+            zlistx_destroy (&keys);
+            keys = NULL;
+        }
+        n = zlistx_next (rl->nodes);
+    }
+
+    return properties;
+error:
+    saved_errno = errno;
+    zhashx_destroy (&properties);
+    if (keys)
+        zlistx_destroy (&keys);
+    errno = saved_errno;
+    return NULL;
+}
+
+static json_t *rlist_json_properties (struct rlist *rl)
+{
+    int saved_errno;
+    zhashx_t *properties = NULL;;
+    json_t *o = NULL;
+    json_t *result = NULL;
+    struct idset *ids;
+
+    if (!(properties = rlist_properties (rl)))
+        return NULL;
+
+    /*  Do not bother returning an empty JSON object
+     */
+    if (zhashx_size (properties) == 0)
+        goto out;
```

Comment:
Do we need `errno = EINVAL`?

Reply:
This isn't an error condition, so I don't think so.